### PR TITLE
Transfer learning

### PR DESCRIPTION
**Pull request by danich1**
This PR is uploading my previous experiment where we tested to see if label functions could transfer between relationship types. Majority of this PR contains files that I added to run this experiment. To save time only concern yourself with the **transfer_learning.ipynb** file. Note: reason for erratic performance when training the models is that I am predicting only 50 sentences. For future reference more data means more stable results.

### CODE REVIEW

File: compound_disease/compound_treats_disease/transfer_learning_experiment/data/label_functions/compound_gene_lf.py [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+        if gene_entry_df["Symbol"].values[0].lower() == token or token in gene_entry_df["Synonyms"].values[0].lower():
+            return 0
+        elif token in gene_entry_df["description"].values[0].lower():
+            return 0
+    return -1
+
+"""
+SENTENCE PATTERN MATCHING
+"""
+
```
Comment (ajlee21) [inline compound_disease/compound_treats_disease/transfer_learning_experiment/data/label_functions/compound_gene_lf.py:182]: Just to clarify, is "binding_indication" the relationship on your edge and the different strings here are the specific pattern you are searching for in the sentence?

Are these patterns generated by you based on reading?  Or is there some source?

File: compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+from snorkel.models import candidate_subclass
+
+from metal.label_model import LabelModel
+
+from compound_disease_lf import CD_LFS
+from disease_gene_lf import DG_LFS
+from compound_gene_lf import CG_LFS
+from utils.notebook_utils.dataframe_helper import load_candidate_dataframes
+from utils.notebook_utils.label_matrix_helper import (
+    get_columns, label_candidates, get_auc_significant_stats
```
Comment (ajlee21) [inline compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py:57]: (for my personal growth as a programmer) Are you importing python NB here?  I didn't realize you could do this?

File: compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+# In[ ]:
+
+
+plot_label_matrix_heatmap(label_matricies['train'].T, 
+                          yaxis_tick_labels=lf_names, 
+                          figsize=(10,13))
+
+
+# # Train Generative Model
+
```
Comment (ajlee21) [inline compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py:151]: What are these different models?  I think I might be misunderstanding how you're using this term.  Looks like these are different hyperparams of your model that you're using

File: compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+for model in tqdm_notebook(model_dict):
+    for reg_param in reg_param_grid:
+        label_model.train(L[:, model_dict[model]], n_epochs=1000, verbose=False, lr=0.01, l2=reg_param)
+        grid_results[model][str(reg_param)] = label_model.predict_proba(L_dev[:, model_dict[model]])[:,0]
+
+
+# In[ ]:
+
+
+for model in grid_results:
```
Comment (ajlee21) [inline compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py:206]: You're dev is your holdout set from your training?  For your dev set you have labels and you're using the performance on the dev set to select your hyperparams?

There's not CV here correct?

**Review by ajlee21 — APPROVED** [review]
Looks pretty good.  Some questions to clarify what you're doing here

File: compound_disease/compound_treats_disease/transfer_learning_experiment/data/label_functions/compound_gene_lf.py (hunk 1 above)
Comment (danich1) [reply compound_disease/compound_treats_disease/transfer_learning_experiment/data/label_functions/compound_gene_lf.py:182]: binding_indication is a set that contains keywords to indicate a compound binding to a gene or vise versa. Majority of these trigger words come from me reading the literature.

File: compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py (hunk 3 above)
Comment (danich1) [reply compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py:151]: The different models are the combination of label functions I am using to train the generative model. For example I have a section where I am using label functions designed for predicting Compound binds Gene (CbG) relationships to label and predict compound disease sentences. The goal here is to see how well label functions can transfer between relationship types.

File: compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py (hunk 4 above)
Comment (danich1) [reply compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py:206]: Correct in this context the training set is unlabeled which means one cannot use CV without significant amount of effort to get gold labels. In this case I am using a dev set to select a model.

**Review by dhimmel — COMMENTED** [review]
I see Alex is also reviewing. If my review would be appreciated, I will stop by David's desk to go over [`transfer_learning.ipynb`](https://github.com/greenelab/snorkeling/blob/4b240466aa3aaea096d3caa0589a9ed5cea19836/compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.ipynb) to make sure I understand everything.

File: compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py (hunk 2 above)
Comment (danich1) [reply compound_disease/compound_treats_disease/transfer_learning_experiment/transfer_learning.py:57]: Not sure you can import notebooks, but in this case I'm just importing scripts that placed in a folder called notebook_utils.

**Review by dhimmel — COMMENTED** [review]
Noting that I discussed PR with David prior to merge.
